File: sdk/mobile/analytiq.swift
// Analytiq iOS/macOS SDK
// Swift 5.9+  |  iOS 13+, macOS 10.15+, tvOS 13+, watchOS 6+
//
// Usage:
//   import Analytiq
//
//   // Initialize once in AppDelegate / @main
//   Analytiq.configure(apiKey: "YOUR_API_KEY", host: "https://your-domain.com")
//
//   // Track an event
//   Analytiq.track("button_tapped", properties: ["screen": "home"])
//
//   // Identify a user
//   Analytiq.identify("user-123", traits: ["email": "alice@example.com", "plan": "pro"])
//
//   // Page view
//   Analytiq.page("Settings")

import Foundation

public final class Analytiq: @unchecked Sendable {

    // ── Configuration ──────────────────────────────────────────────────────

    private static var shared: Analytiq?

    private let apiKey: String
    private let host:   String
    private let queue  = DispatchQueue(label: "com.analytiq.sdk", qos: .utility)
    private let session: URLSession

    private var userId:      String?
    private var anonymousId: String

    private init(apiKey: String, host: String) {
        self.apiKey   = apiKey
        self.host     = host.hasSuffix("/") ? String(host.dropLast()) : host
        let cfg       = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest  = 10
        cfg.timeoutIntervalForResource = 30
        self.session  = URLSession(configuration: cfg)

        // Stable anonymous ID persisted across launches
        let anonKey = "analytiq_anon_id"
        if let stored = UserDefaults.standard.string(forKey: anonKey) {
            self.anonymousId = stored
        } else {
            let id = UUID().uuidString
            UserDefaults.standard.set(id, forKey: anonKey)
            self.anonymousId = id
        }
    }

    // ── Public API ─────────────────────────────────────────────────────────

    public static func configure(apiKey: String, host: String = "https://app.analytiq.io") {
        shared = Analytiq(apiKey: apiKey, host: host)
    }

    public static func identify(_ userId: String, traits: [String: Any] = [:]) {
        guard let sdk = shared else { return }
        sdk.queue.async {
            sdk.userId = userId
            sdk.send(type: "identify", event: nil, userId: userId, properties: traits)
        }
    }

    public static func track(_ event: String, properties: [String: Any] = [:]) {
        guard let sdk = shared else { return }
        sdk.queue.async {
            sdk.send(type: "track", event: event, userId: sdk.userId, properties: properties)
        }
    }

    public static func page(_ name: String, properties: [String: Any] = [:]) {
        guard let sdk = shared else { return }
        sdk.queue.async {
            var props = properties
            props["name"] = name
            sdk.send(type: "page", event: name, userId: sdk.userId, properties: props)
        }
    }

    public static func reset() {
        guard let sdk = shared else { return }
        sdk.queue.async {
            sdk.userId = nil
            let id = UUID().uuidString
            sdk.anonymousId = id
            UserDefaults.standard.set(id, forKey: "analytiq_anon_id")
        }
    }

    // ── Internal ───────────────────────────────────────────────────────────

    private func send(type: String, event: String?, userId: String?, properties: [String: Any]) {
        var body: [String: Any] = [
            "type":        type,
            "anonymousId": anonymousId,
            "properties":  properties,
        ]
        if let event  = event  { body["event"]  = event }
        if let userId = userId { body["userId"] = userId }

        guard
            let url  = URL(string: "\(host)/api/ingest/\(apiKey)"),
            let data = try? JSONSerialization.data(withJSONObject: body)
        else { return }

        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.httpBody   = data
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        session.dataTask(with: req) { _, _, error in
            if let error = error {
                print("[Analytiq] ingest error: \(error.localizedDescription)")
            }
        }.resume()
    }
}

File: sdk/mobile/Analytiq.kt
// Analytiq Android SDK
// Kotlin 1.9+  |  Android API 21+ (Android 5.0+)
//
// Usage:
//   // Initialize once in Application.onCreate()
//   Analytiq.configure(context, apiKey = "YOUR_API_KEY", host = "https://your-domain.com")
//
//   // Track
//   Analytiq.track("button_tapped", mapOf("screen" to "home"))
//
//   // Identify
//   Analytiq.identify("user-123", mapOf("email" to "alice@example.com", "plan" to "pro"))
//
//   // Page/screen
//   Analytiq.screen("Settings")

package io.analytiq.sdk

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

object Analytiq {

    private const val PREFS_NAME = "analytiq_prefs"
    private const val ANON_KEY   = "anonymous_id"

    private var apiKey:      String = ""
    private var host:        String = "https://app.analytiq.io"
    private var userId:      String? = null
    private var anonymousId: String = UUID.randomUUID().toString()

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // ── Configuration ──────────────────────────────────────────────────────

    fun configure(context: Context, apiKey: String, host: String = "https://app.analytiq.io") {
        this.apiKey = apiKey
        this.host   = host.trimEnd('/')

        val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        anonymousId = prefs.getString(ANON_KEY, null) ?: run {
            val id = UUID.randomUUID().toString()
            prefs.edit().putString(ANON_KEY, id).apply()
            id
        }
    }

    // ── Public API ─────────────────────────────────────────────────────────

    fun identify(userId: String, traits: Map<String, Any> = emptyMap()) {
        this.userId = userId
        enqueue("identify", null, userId, traits)
    }

    fun track(event: String, properties: Map<String, Any> = emptyMap()) {
        enqueue("track", event, userId, properties)
    }

    fun screen(name: String, properties: Map<String, Any> = emptyMap()) {
        val props = properties.toMutableMap().apply { put("name", name) }
        enqueue("page", name, userId, props)
    }

    fun reset() {
        userId = null
    }

    // ── Internal ───────────────────────────────────────────────────────────

    private fun enqueue(
        type:       String,
        event:      String?,
        userId:     String?,
        properties: Map<String, Any>,
    ) {
        scope.launch {
            try {
                val body = JSONObject().apply {
                    put("type",        type)
                    put("anonymousId", anonymousId)
                    put("properties",  JSONObject(properties))
                    event?.let  { put("event",  it) }
                    userId?.let { put("userId", it) }
                }
                post("$host/api/ingest/$apiKey", body.toString())
            } catch (e: Exception) {
                System.err.println("[Analytiq] ingest error: ${e.message}")
            }
        }
    }

    private fun post(url: String, json: String) {
        val conn = URL(url).openConnection() as HttpURLConnection
        try {
            conn.requestMethod = "POST"
            conn.setRequestProperty("Content-Type", "application/json")
            conn.doOutput = true
            conn.connectTimeout = 10_000
            conn.readTimeout    = 10_000
            OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(json) }
            conn.responseCode  // trigger the request
        } finally {
            conn.disconnect()
        }
    }
}
